### PR TITLE
 add no-commented-code rule

### LINKED ISSUE #17: Rule: Sections of code should not be "commented out"
[RSPEC-125](https://jira.sonarsource.com/browse/RSPEC-125)

### PR DESCRIPTION
**Pull request by stas-vilchik**
Fixes #17

### CODE REVIEW

File: sonarts-core/src/utils/sonar-analysis.ts [hunk 1]
```diff
@@ -82,20 +86,17 @@ export class IssueLocation {
...
   public constructor(primaryLocation: IssueLocation, ruleName: string) {
-    super(
-      primaryLocation.node.getSourceFile(),
-      primaryLocation.node.getStart(),
-      primaryLocation.lastNode.getEnd(),
-      primaryLocation.message!,
-      ruleName,
-    );
-
+    super(primaryLocation.sourceFile, primaryLocation.start, primaryLocation.end, primaryLocation.message!, ruleName);
```
Comment (SonarTech) [inline sonarts-core/src/utils/sonar-analysis.ts]: Bug Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=stas%2Fno-commented-code&issues=AWC7g8QeqDrCo5ysghkQ&open=AWC7g8QeqDrCo5ysghkQ)

[ISSUE_KEY]: AWC7g8QeqDrCo5ysghkQ

File: sonarts-core/src/rules/noCommentedCodeRule.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  sourceFile: ts.SourceFile;
+
+  protected visitSourceFile(sourceFile: ts.SourceFile) {
+    this.sourceFile = sourceFile;
+    const tokens = toTokens(sourceFile);
+    tokens.forEach(token => {
+      this.processComments(getCommentsBefore(token), sourceFile);
+      this.processComments(getCommentsAfter(token), sourceFile);
+    });
+    super.visitSourceFile(sourceFile);
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noCommentedCodeRule.ts]: @stas-vilchik we can drop this line

File: sonarts-core/src/rules/noCommentedCodeRule.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  }
+
+  private processComments(comments: ts.CommentRange[], sourceFile: ts.SourceFile) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noCommentedCodeRule.ts]: @stas-vilchik why passing source file if you have property?

File: sonarts-core/src/rules/noCommentedCodeRule.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        currentGroup = [comment];
+      } else if (this.isAdjacent(comment, currentGroup)) {
+        currentGroup.push(comment);
+      } else {
+        groups.push(currentGroup);
+        currentGroup = [comment];
+      }
+    }
+
+    if (currentGroup != null) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noCommentedCodeRule.ts]: @stas-vilchik can't we omit `!= null`?

File: sonarts-core/src/rules/noCommentedCodeRule.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+      return comment.substring(2);
+    } else if (comment.startsWith("/*")) {
+      return comment.substring(2, comment.length - 2);
+    } else if (comment.startsWith("<!--")) {
+      if (comment.endsWith("-->")) {
+        return comment.substring(4, comment.length - 3);
+      }
+      return comment.substring(4);
+    } else {
+      throw new Error();
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noCommentedCodeRule.ts]: @stas-vilchik some error message would be good here

**Comment by vilchik-elena** [comment]
Manually tested

**Pull request merged**